# Signature_Validation

Functional gene expression signatures (FGESs) are commonly used in tumor transcriptomic analysis, including predictive and prognostic models. However, FGES creation and validation is still a challenge due to both technical and biological limitations, resulting in no well-established criteria to assess FGES quality. Here, we describe a comprehensive pipeline to develop and validate FGESs, which includes the following steps: creation of a gene set through both computational and analytical methods; calculation of the gene set score by a modified single-sample gene set enrichment analysis (ssGSEA) and its quality assessment according to a defined set of technical and biological criteria, which ensures reproducibility, specificity, and robustness; FGES refinement and final validation. 20 cell-type-specific and 5 process-describing FGES created according to the proposed workflow demonstrated higher biological relevance compared to most publicly available gene signatures. Thus, the proposed workflow improves control of the FGES quality for their further use in investigation of complex biological processes, including analysis of tumor microenvironment.
